src/common/Cryptography/SHA1.h

```diff
@@ -29,21 +30,36 @@ class TC_COMMON_API SHA1Hash
     public:
         SHA1Hash();
         ~SHA1Hash();
+        static constexpr size_t HASH_LEN = SHA_DIGEST_LENGTH;
 
-        void UpdateBigNumbers(BigNumber* bn0, ...);
 
         void UpdateData(const uint8 *dta, int len);
         void UpdateData(const std::string &str);
 
+        template <size_t SIZE, size_t... SIZES, typename... Ts> void UpdateBigNumbers(BigNumber const& bn, Ts&&... tail)
+        {
+            static_assert(sizeof...(SIZES) == sizeof...(Ts));
+            UpdateBigNumber<SIZE>(bn);
+            if constexpr (sizeof...(Ts) > 0)
+                UpdateBigNumbers<SIZES...>(std::forward<Ts>(tail)...);
+        }
+
```

Comment:
```cpp
    template <
        typename... Ts>
    auto UpdateBigNumbers(Ts&&... ts) -> std::enable_if_t<
        (std::is_same_v<foo, std::decay_t<Ts>> && ...)
    , void>
    {
        return (UpdateBigNumber(std::forward<Ts>(ts)), ...);
    }
```

You can tweak the decay bit, i was feeling lazy

Reply:
asserting can produce a nicer error message though

Reply:
sure. fold expressions though

Reply:
ha sure we can do it